controllers/webhook/runcompletion_feed.go

```diff
@@ -64,52 +67,115 @@ func getRequestBody(ctx context.Context, request *http.Request) ([]byte, error)
+			if k8sErrors.IsNotFound(err) {
+				return nil, &MissingResourceError{err.Error()}
+			}
+			return nil, &FatalError{err.Error()}
+		}
+		return rc, nil
 	}
+
+	return nil, nil
+}
+
+func (rcf RunCompletionFeed) fetchRun(ctx context.Context, name *common.NamespacedName) (*pipelineshub.Run, EventError) {
+	logger := common.LoggerFromContext(ctx)
+	if name != nil {
+		run := &pipelineshub.Run{}
+		if err := rcf.client.Get(ctx, client.ObjectKey{
+			Namespace: name.Namespace,
+			Name:      name.Name,
+		}, run); err != nil {
+			logger.Error(err, "failed to load", "Run", name)
+			if k8sErrors.IsNotFound(err) {
+				return nil, &MissingResourceError{err.Error()}
+			}
+			return nil, &FatalError{err.Error()}
+		}
+		return run, nil
+	}
+
+	return nil, nil
 }
 
 func (rcf RunCompletionFeed) handleEvent(response http.ResponseWriter, request *http.Request) {
 	logger := log.FromContext(rcf.ctx)
 	switch request.Method {
 	case http.MethodPost:
+
 		if request.Header.Get(HttpHeaderContentType) != HttpContentTypeJSON {
 			logger.Error(errors.New("RunCompletionFeed call failed"), fmt.Sprintf("invalid %s [%s], want `%s`", HttpHeaderContentType, request.Header.Get(HttpHeaderContentType), HttpContentTypeJSON))
 			http.Error(response, fmt.Sprintf("invalid %s, want `%s`", HttpHeaderContentType, HttpContentTypeJSON), http.StatusUnsupportedMediaType)
 			return
 		}
-		event, err := rcf.extractRunCompletionEvent(request)
-		if err != nil || event == nil {
-			logger.Error(err, "Failed to extract body from request")
-			http.Error(response, err.Error(), http.StatusBadRequest)
+
+		eventData, err := rcf.extractRunCompletionEventData(request)
+		if err != nil {
+			err.ToHttpError(response)
 			return
-		} else {
-			for _, handler := range rcf.eventHandlers {
-				err := handler.Handle(*event)
-				if err != nil {
-					logger.Error(err, "Run completion event handler operation failed")
+		}
+
+		var runConfiguration *pipelineshub.RunConfiguration
+		if eventData.RunConfigurationName != nil || eventData.RunConfigurationName.Name != "" {
+			runConfiguration, err = rcf.fetchRunConfiguration(rcf.ctx, eventData.RunConfigurationName)
+			if err != nil {
+				err.ToHttpError(response)
+				return
+			}
+		}
+
+		var run *pipelineshub.Run
+		if eventData.RunName != nil || eventData.RunName.Name != "" {
+			run, err = rcf.fetchRun(rcf.ctx, eventData.RunName)
+			if err != nil {
+				err.ToHttpError(response)
+				return
+			}
+		}
+
+		event, err := rcf.eventProcessor.ToRunCompletionEvent(eventData, runConfiguration, run)
+		if err != nil {
+			err.ToHttpError(response)
+			return
+		}
+
+		for _, handler := range rcf.eventHandlers {
+			err := handler.Handle(*event)
+			if err != nil {
+				logger.Error(err, "Run completion event handler operation failed")
+				if k8sErrors.IsNotFound(err) {
+					http.Error(response, err.Error(), http.StatusGone)
+					return
+				} else {
 					http.Error(response, err.Error(), http.StatusInternalServerError)
```

Comment:
Here we return `http.StatusInternalServerError`, but none of the error type currently defined in `errors.go` returns such an error, so we probably need to add one to cover this case.